lib/cc/services/stash_pull_requests.rb

```diff
@@ -0,0 +1,98 @@
+require "base64"
+require "cc/presenters/pull_requests_presenter"
+
+class CC::Service::StashPullRequests < CC::Service
+  class Config < CC::Service::Config
+    attribute :domain, String,
+      description: "Your Stash host domain (e.g. yourstash.com:PORT, please exclude https://)"
+    attribute :username, String
+    attribute :password, Password
+
+    validates :domain, presence: true
+    validates :username, presence: true
+    validates :password, presence: true
+  end
+
+  self.title = "Stash Pull Requests"
+  self.description = "Update pull requests on Stash"
+
+  STASH_STATES = {
+    "error" => "FAILED",
+    "failure" => "FAILED",
+    "pending" => "INPROGRESS",
+    "skipped" => "SUCCESSFUL",
+    "success" => "SUCCESSFUL",
+  }.freeze
+
+  def receive_test
+    setup_http
+
+    service_get(test_url)
+
+    { ok: true, message: "Test succeeded" }
+  rescue HTTPError => e
+    { ok: false, message: e.message }
+  end
+
+  def receive_pull_request
+    setup_http
+
+    params = {
+      description: description,
+      key: "Code Climate",
+      name: "Code Climate",
+      state: state,
+      url:  @payload["details_url"],
+    }
+    service_post(url, params.to_json)
+  end
+
+  private
+
+  def test_url
+    "http://#{config.domain}/rest/api/1.0/users"
+  end
+
+  def url
+    "http://#{config.domain}/rest/build-status/1.0/commits/#{commit_sha}"
+  end
+
+  def commit_sha
+    @payload.fetch("commit_sha")
+  end
+
+  def description
+    return @payload["message"] if @payload["message"]
+
+    case @payload["state"]
+    when "pending"
+      "Code Climate is analyzing this code."
```

Comment:
These message descriptions seem like prime candidates for DRYing across the two kinds of PRs.